superduperdb/ext/llm/model.py

```diff
@@ -0,0 +1,398 @@
+        super().__post_init__()
+
+    def init_model_and_tokenizer(self):
+        model = AutoModelForCausalLM.from_pretrained(
+            self.model_name_or_path,
+            **self.model_kwargs,
+        )
+
+        tokenizer = AutoTokenizer.from_pretrained(
+            self.model_name_or_path,
+            **self.tokenizer_kwags,
+        )
+        tokenizer.pad_token = tokenizer.pad_token or tokenizer.unk_token
+        return model, tokenizer
+
+    def create_trainer(
+        self, train_dataset, eval_dataset, training_args, **kwargs
+    ) -> transformers.Trainer:
+        trainer = transformers.Trainer(
+            model=self.model,
+            tokenizer=self.tokenizer,
+            args=training_args,
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            **kwargs,
+        )
+        return trainer
+
+    def init(self):
+        self.model, self.tokenizer = self.init_model_and_tokenizer()
+
+    def _fit(
+        self,
+        X: t.Any,
+        y: t.Optional[t.Any] = None,
+        configuration: t.Optional[_TrainingConfiguration] = None,
+        data_prefetch: bool = False,
+        db: t.Optional["Datalayer"] = None,
+        metrics: t.Optional[t.Sequence["Metric"]] = None,
+        select: t.Optional["Select"] = None,
+        validation_sets: t.Optional[t.Sequence[t.Union[str, "Dataset"]]] = None,
+        **kwargs,
+    ):
+        assert configuration is not None, "configuration must be provided"
+
+        training_args = LLMTrainingArguments(**configuration.kwargs)  # type: ignore
+
+        # get device map
+        device_map: t.Union[None, str, t.Dict[str, int]] = None
+        if os.environ.get("LOCAL_RANK") is not None:
+            device_map = {"": int(os.environ.get("LOCAL_RANK", "0"))}
+        elif torch.backends.mps.is_available():
+            device_map = "mps"
+
+        quantization_config = self._create_quantization_config(training_args)
+
+        self.model_kwargs["quantization_config"] = quantization_config
+        self.model_kwargs["device_map"] = device_map
+        self.model, self.tokenizer = self.init_model_and_tokenizer()
+
+        self.tokenizer.model_max_length = (
+            training_args.max_length or self.tokenizer.model_max_length
+        )
+        self._prepare_lora_training(training_args)
+
+        train_dataset, eval_dataset = self.get_datasets(
+            X,
+            y,
+            db,
+            select,
+            data_prefetch=data_prefetch,
+            eval=True,
+            prefetch_size=kwargs.pop("prefetch_size", 10000),
+        )
+
+        # TODO: Defind callbacks about superduperdb side
+        trainer = self.create_trainer(
+            train_dataset,
+            eval_dataset,
+            compute_metrics=metrics,
+            training_args=training_args,
+            **kwargs,
+        )
+        trainer.model.config.use_cache = False
+        trainer.train()
+        trainer.save_state()
```

Comment:
Is it possible to save the state during training?

Reply:
It will also be saved during training. This operation is saved again after the training is completed.